### PR TITLE
added User authentication and sessions

### PR DESCRIPTION
**Pull request by tubhutani13**

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -43,7 +43,7 @@ gem "tzinfo-data", platforms: %i[ mingw mswin x64_mingw jruby ]
 gem "bootsnap", require: false
 
 # Use Sass to process CSS
-# gem "sassc-rails"
+gem "sassc-rails"
```
Comment (shubhamjuneja1) [inline Gemfile]: version?

File: app/controllers/home_controller.rb [hunk 2]
```diff
@@ -0,0 +1,2 @@
+class HomeController < ApplicationController
```
Comment (shubhamjuneja1) [inline app/controllers/home_controller.rb]: are we using this?

File: app/controllers/password_resets_controller.rb [hunk 3]
```diff
@@ -0,0 +1,44 @@
+class PasswordResetsController < ApplicationController
```
Comment (shubhamjuneja1) [inline app/controllers/password_resets_controller.rb]: renam to passwordscontroller

File: app/controllers/password_resets_controller.rb [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  before_action :set_user_by_password_reset_token, only: [:edit, :update]
+  before_action :set_user_by_email, only: [:create]
+
+  def new
+  end
+
+  def create
+    if @user
```
Comment (shubhamjuneja1) [inline app/controllers/password_resets_controller.rb]: redirect in before_action, no need to handle here

File: app/controllers/password_resets_controller.rb [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+      @user.send_password_reset
+      flash[:notice] = t("E-mail sent with password reset instructions.")
```
Comment (shubhamjuneja1) [inline app/controllers/password_resets_controller.rb]: is this i18?

File: app/controllers/application_controller.rb [hunk 6]
```diff
@@ -1,2 +1,13 @@
 class ApplicationController < ActionController::Base
+  include SessionsHelper
+
+  private
+
+  def logged_in_user
```
Comment (shubhamjuneja1) [inline app/controllers/application_controller.rb]: rename

File: app/helpers/sessions_helper.rb [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    redirect_to(session[:forwarding_url] || default)
+    session.delete(:forwarding_url)
+  end
+
+  def store_location
+    session[:forwarding_url] = request.original_url if request.get?
+  end
+
+  private def cookie_expiration_time
+    params[:remember_me] == '1' ? 1.day.from_now : nil
```
Comment (shubhamjuneja1) [inline app/helpers/sessions_helper.rb]: use constant instead of 1.day

File: app/controllers/application_controller.rb [hunk 8]
```diff
@@ -1,2 +1,13 @@
...
+    unless logged_in?
+      store_location
+      flash[:danger] = "Please log in."
```
Comment (shubhamjuneja1) [inline app/controllers/application_controller.rb]: i18

File: app/helpers/sessions_helper.rb [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+  def current_user
+    @current_user ||= User.find_by(id: cookies.encrypted[:user_id])
+  end
+
+  def logged_in?
+    current_user.present?
+  end
+
+  def log_out
+    @current_user = nil
```
Comment (shubhamjuneja1) [inline app/helpers/sessions_helper.rb]: why this needed?

File: app/controllers/application_controller.rb [hunk 10]
```diff
@@ -1,2 +1,13 @@
 class ApplicationController < ActionController::Base
+  include SessionsHelper
+
+  private
+
+  def logged_in_user
+    unless logged_in?
+      store_location
```
Comment (shubhamjuneja1) [inline app/controllers/application_controller.rb]: please show if this is working fine

File: app/controllers/password_resets_controller.rb (hunk 3 above)
Comment (shubhamjuneja1) [reply app/controllers/password_resets_controller.rb]: not done yet

File: app/controllers/password_resets_controller.rb [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+  end
+
+  private
+
+  def user_params
+    params.require(:user).permit(:password)
+  end
+
+  def set_user_by_password_reset_token
+    @user = User.find_by_password_reset_token!(params[:id])
```
Comment (shubhamjuneja1) [inline app/controllers/password_resets_controller.rb]: redirect from here

File: app/controllers/password_resets_controller.rb [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+  end
+
+  def set_user_by_email
+    @user = User.find_by_email(params[:email])
```
Comment (shubhamjuneja1) [inline app/controllers/password_resets_controller.rb]: redirect from here

File: app/controllers/password_resets_controller.rb [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+      flash[:notice] = t("E-mail sent with password reset instructions.")
+      redirect_to new_session_path
+    else
+      flash[:error] = "Invalid email address"
+      render :new, status: :unprocessable_entity
+    end
+  end
+
+  def update
+    if @user.password_reset_sent_at < 2.hour.ago
```
Comment (shubhamjuneja1) [inline app/controllers/password_resets_controller.rb]: should be only on model level

File: app/controllers/sessions_controller.rb [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+  before_action :authenticate_user, only: [:create]
+
+  def create
+    if @user.verified?
+      log_in @user
+      redirect_to root_url
+    else
+      flash.now[:error] = "Please activate your account by following the 
+        instructions in the account confirmation email you received to proceed"
+      render "new", status: :unprocessable_entity
```
Comment (shubhamjuneja1) [inline app/controllers/sessions_controller.rb]: why rendering and not redirecting here?

File: app/controllers/sessions_controller.rb [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+  end
+
+  def destroy
+    log_out
+    redirect_to root_url
+  end
+
+  private def set_user
+    unless @user = User.find_by(email: params[:email].downcase)
+      redirect_to login_url, notice: t("invalid_combination")
```
Comment (shubhamjuneja1) [inline app/controllers/sessions_controller.rb]: rename invalid_combination

File: app/controllers/users_controller.rb [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+  before_action :set_user_by_email_confirm_token, only: [:confirm_email]
+
+  def new
+    @user = User.new
+  end
+
+  def create
+    @user = User.new(user_params)
+    if @user.save
+      UserMailer.registration_confirmation(@user.id).deliver
```
Comment (shubhamjuneja1) [inline app/controllers/users_controller.rb]: should be in callback

File: app/controllers/users_controller.rb [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+      flash[:success] = "Please confirm your email address to continue"
```
Comment (shubhamjuneja1) [inline app/controllers/users_controller.rb]: handle i18 everywhere

File: app/controllers/users_controller.rb [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+      redirect_to root_url
+    else
+      flash[:error] = "Ooooppss, something went wrong!"
+      render :new, status: :unprocessable_entity
+    end
+  end
+
+  def confirm_email
+    if @user
+      @user.email_activate
```
Comment (shubhamjuneja1) [inline app/controllers/users_controller.rb]: what if this failed?

File: app/controllers/users_controller.rb [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+      flash[:success] = "Please confirm your email address to continue"
+      redirect_to root_url
+    else
+      flash[:error] = "Ooooppss, something went wrong!"
+      render :new, status: :unprocessable_entity
+    end
+  end
+
+  def confirm_email
+    if @user
```
Comment (shubhamjuneja1) [inline app/controllers/users_controller.rb]: before_action

File: app/mailers/user_mailer.rb [hunk 20]
```diff
@@ -0,0 +1,15 @@
...
+  default :from => DEFAULT_EMAIL
+
+  def registration_confirmation(user_id)
+    @user = User.find(user_id)
+    mail(:to => "#{@user.name} <#{@user.email}>", :subject => t("registration_confirmation"))
+  end
+
+  def forgot_password(user_id)
+    @user = User.find(user_id)
+    @greeting = "Hi"
```
Comment (shubhamjuneja1) [inline app/mailers/user_mailer.rb]: no need of variable

File: app/models/user.rb [hunk 21]
```diff
@@ -0,0 +1,51 @@
+class User < ApplicationRecord
+  enum role: {
+         "admin" => 1,
```
Comment (shubhamjuneja1) [inline app/models/user.rb]: extra spaces

Comment (shubhamjuneja1) [reply app/models/user.rb]: try to keep enum in sequence, easier to add more

File: app/models/user.rb [hunk 22]
```diff
@@ -0,0 +1,51 @@
...
+         "user" => 0,
+       }
+
+  has_secure_password
+  before_create :confirmation_token
+  before_save { self.email = email.downcase }
```
Comment (shubhamjuneja1) [inline app/models/user.rb]: create method

File: app/models/user.rb [hunk 23]
```diff
@@ -0,0 +1,51 @@
+class User < ApplicationRecord
+  enum role: {
+         "admin" => 1,
+         "user" => 0,
+       }
+
+  has_secure_password
+  before_create :confirmation_token
```
Comment (shubhamjuneja1) [inline app/models/user.rb]: rename

File: app/models/user.rb [hunk 24]
```diff
@@ -0,0 +1,51 @@
...
+  before_save { self.email = email.downcase }
+  validates :name, presence: true
+  validates :email, presence: true, format: { with: VALID_EMAIL_REGEX }, uniqueness: { case_sensitive: false }
+  validates :password, presence: true, length: { minimum: 6 }, if: :password_set?
+
+  def verified?
+    if self.verified_at.nil?
```
Comment (shubhamjuneja1) [inline app/models/user.rb]: this can be simplified

File: app/models/user.rb [hunk 25]
```diff
@@ -0,0 +1,51 @@
...
+      return false
+    end
+    true
+  end
+
+  def email_activate
+    self.verified_at = Time.now
```
Comment (shubhamjuneja1) [inline app/models/user.rb]: can't we simply use
update(email_confirm_token: nil, verified_at: Time.current)

File: app/models/user.rb [hunk 26]
```diff
@@ -0,0 +1,51 @@
...
+    self.email_confirm_token = nil
+    save!
```
Comment (shubhamjuneja1) [inline app/models/user.rb]: why save! and not save

File: app/models/user.rb [hunk 27]
```diff
@@ -0,0 +1,51 @@
...
+
+  private
+
+  def confirmation_token
+    if self.email_confirm_token.blank?
+      self.email_confirm_token = SecureRandom.urlsafe_base64.to_s
+    end
+  end
+
+  def generate_token(column)
```
Comment (shubhamjuneja1) [inline app/models/user.rb]: this should not be a method in user model
instead this should be a module and call the method here

File: app/models/user.rb [hunk 28]
```diff
@@ -0,0 +1,51 @@
...
+  end
+
+  def email_activate
+    self.verified_at = Time.now
+    self.email_confirm_token = nil
+    save!
+  end
+
+  def send_password_reset
+    generate_token(:password_reset_token)
```
Comment (shubhamjuneja1) [inline app/models/user.rb]: if update(token: TokenHandler.generate_token, password_reset_sent_at: Time.now)
 UserMailer.forgot_password(self.id).deliver
end

File: test/fixtures/users.yml [hunk 29]
```diff
@@ -0,0 +1,21 @@
+# Read about fixtures at https://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
+
+one:
+  name: MyString
+  email: MyString
+  password_digest: MyString
+  role: 1
+  verified_at: 2023-01-19 16:42:43
+  email_confirm_token: MyString
```
Comment (shubhamjuneja1) [inline test/fixtures/users.yml]: what's this?

**Review by shubhamjuneja1 — CHANGES_REQUESTED** [review]

File: app/controllers/home_controller.rb (hunk 2 above)
Comment (tubhutani13) [reply app/controllers/home_controller.rb]: yes. to show the index

File: app/helpers/sessions_helper.rb (hunk 9 above)
Comment (tubhutani13) [reply app/helpers/sessions_helper.rb]: when we are destroying the session then we are logging out

File: app/controllers/sessions_controller.rb (hunk 14 above)
Comment (tubhutani13) [reply app/controllers/sessions_controller.rb]: Sir, the user is already at the login page it is there when the user is trying to login but is not verified

File: app/controllers/password_resets_controller.rb (hunk 11 above)
Comment (tubhutani13) [reply app/controllers/password_resets_controller.rb]: please discuss

File: app/controllers/password_resets_controller.rb (hunk 12 above)
Comment (tubhutani13) [reply app/controllers/password_resets_controller.rb]: please discuss

File: test/fixtures/users.yml (hunk 29 above)
Comment (tubhutani13) [reply test/fixtures/users.yml]: auto generated

File: app/controllers/password_resets_controller.rb (hunk 13 above)
Comment (tubhutani13) [reply app/controllers/password_resets_controller.rb]: discuss

File: app/models/user.rb (hunk 27 above)
Comment (tubhutani13) [reply app/models/user.rb]: discuss

**Pull request merged**